### PR TITLE
ast: brand new day with a brand new branch

### PR DESCRIPTION
**Pull request by vlad-scherbich**
This `ast.mli` is a copy of the one from https://github.com/Howon/Oscar/pull/12 plus:
- defined `actor_receive`
- defined `function_formals`

### CODE REVIEW

File: src/ast.mli [hunk 1]
```diff
@@ -1,83 +1,91 @@
+type program = None
+
 type bin_op =
-    | Add    | Minus | Multiply | Divide
-    | Modulo | Equal | Neq      | Lt
-    | Leq    | Gt    | Geq      | And
-    | Or
+  Add    | Minus | Multiply | Divide
```
Comment (vlad-scherbich) [inline src/ast.mli]: So I'll remove the other leading `|`s in type definitions as they're not needed. LMK if anyone objects.

Comment (Howon) [reply src/ast.mli]: thats fine

File: src/ast.mli [hunk 2]
```diff
@@ -1,83 +1,91 @@
...
-    function_body : stmt list;
-}
-
 type actor = {
-    actor_fields : field list;
-    actor_name : string;
-    acator_receive: receive;
+  actor_fields : field list;
+  actor_name : string;
+  (* todo: needs defining *)
```
Comment (Howon) [inline src/ast.mli]: functions can go in here I think

Comment (vlad-scherbich) [reply src/ast.mli]: I see, but the syntax is a little different for functions and receive:

```
def printMaybe(x: maybe<int>) => () = {
  if (x.isDefined()) {
    println(x.get());
  }
}
```

vs

```
let receive => {
    | work(start: int, numElems: int) => {
        result(calcPi(start, numElems)) |> sender 
   }
}
```

File: src/ast.mli [hunk 3]
```diff
@@ -1,83 +1,91 @@
...
-    | Bit_Xor
-    | Bit_Not
+    Big_RShift
+  | Bit_LShift
+  | Bit_And
+  | Bit_Or
+  | Bit_Xor
+  | Bit_Not
+
+type message_type = {
```
Comment (vlad-scherbich) [inline src/ast.mli]: btw, I had to chain these three types b/c of a cycle:
`message_type -> stmt -> expr -> message_type`

File: src/ast.mli [hunk 4]
```diff
@@ -1,83 +1,91 @@
+type program = None
+
 type bin_op =
-    | Add    | Minus | Multiply | Divide
-    | Modulo | Equal | Neq      | Lt
-    | Leq    | Gt    | Geq      | And
-    | Or
+    Add | Sub | Mult | Div
+  | Mod | Equal | Neq | Less
```
Comment (Howon) [inline src/ast.mli]: thank you

File: src/ast.mli (hunk 3 above)
Comment (Howon) [reply src/ast.mli]: sgtm

**Comment by Howon** [comment]
Can you remove AST and OSCAR.ML from this pr?

**Comment by vlad-scherbich** [comment]
I moved scanner/parser changes into a separate PR instead. Keeping AST and Oscar here.

**Comment by Howon** [comment]
why oscar.ml?

**Comment by vlad-scherbich** [comment]
Seems like a nice separation, following HW1 pattern. Do you prefer to keep all the AST code in one file?

**Comment by Howon** [comment]
yes please

**Review by aethos — APPROVED** [review]

**Pull request merged**